xbmc/addons/AddonRepos.cpp

```diff
@@ -0,0 +1,215 @@
+{
+  return IsFromOfficialRepo(addon, false);
+}
+
+bool CAddonRepos::IsFromOfficialRepo(const AddonPtr& addon, bool bCheckAddonPath)
+{
+  auto comparator = [&](const RepoInfo& officialRepo) {
+    if (bCheckAddonPath)
+    {
+      return (addon->Origin() == officialRepo.m_repoId &&
+              StringUtils::StartsWithNoCase(addon->Path(), officialRepo.m_origin));
+    }
+
+    return addon->Origin() == officialRepo.m_repoId;
+  };
+
+  return addon->Origin() == ORIGIN_SYSTEM ||
+         std::any_of(officialRepoInfos.begin(), officialRepoInfos.end(), comparator);
+}
+
+const std::map<std::string, AddonPtr>& CAddonRepos::GetLatestPrivateVersions() const
+{
+  return m_latestPrivateVersions;
+}
+
+const std::map<std::string, AddonPtr>& CAddonRepos::GetLatestOfficialVersions() const
+{
+  return m_latestOfficialVersions;
+}
+
+const std::map<std::string, std::map<std::string, AddonPtr>>& CAddonRepos::GetLatestVersionsByRepo()
+    const
+{
+  return m_latestVersionsByRepo;
+}
+
+void CAddonRepos::SetupFullAddonsByRepoMap(const VECADDONS& addons)
+{
+  m_addonsByRepoMap.clear();
+
+  for (const auto& addon : addons)
+    if (m_addonMgr.IsCompatible(*addon))
+    {
+      m_addonsByRepoMap[addon->Origin()].insert({addon->ID(), addon});
+    }
+
+  for (const auto& map : m_addonsByRepoMap)
+    CLog::Log(LOGDEBUG, "Setting up map for repository: {} - addon count : {}", map.first,
+              map.second.size());
+
+  SetupLatestVersionMaps();
+}
+
+void CAddonRepos::SetupLatestVersionMaps()
+{
+  m_latestOfficialVersions.clear();
+  m_latestPrivateVersions.clear();
+  m_latestVersionsByRepo.clear();
+
+  const AddonRepoUpdateMode updateMode =
+      CAddonSystemSettings::GetInstance().GetAddonRepoUpdateMode();
+
+  for (const auto& repo : m_addonsByRepoMap)
+  {
+    const auto& addonsPerRepo = repo.second;
+
+    for (const auto& addonMapEntry : addonsPerRepo)
+    {
+      const auto& addonToAdd = addonMapEntry.second;
+
+      if (updateMode == AddonRepoUpdateMode::OFFICIAL_ONLY)
+      {
+        if (IsFromOfficialRepo(addonToAdd, true))
+        {
+          AddAddonIfLatest(addonToAdd, m_latestOfficialVersions);
+        }
+        else
+        {
+          AddAddonIfLatest(addonToAdd, m_latestPrivateVersions);
+        }
+      }
+      else if (updateMode == AddonRepoUpdateMode::ANY_REPOSITORY)
+      {
+        AddAddonIfLatest(addonToAdd, m_latestOfficialVersions);
+      }
+
+      // add to latestVersionsByRepo - map
+      AddAddonIfLatest(addonToAdd, repo.first, m_latestVersionsByRepo);
+    }
+  }
+}
+
+void CAddonRepos::AddAddonIfLatest(const AddonPtr& addonToAdd,
+                                   std::map<std::string, AddonPtr>& map) const
+{
+  const auto& latestKnown = map.find(addonToAdd->ID());
+  if (latestKnown == map.end() || addonToAdd->Version() > latestKnown->second->Version())
+    map[addonToAdd->ID()] = addonToAdd;
+}
+
+void CAddonRepos::AddAddonIfLatest(
```

Comment:
I think the addon ID and the repo ID need to be swapped in this function.

It should give you the latest versions of addons per repo. Or have I misunderstood?

Reply:
ready now